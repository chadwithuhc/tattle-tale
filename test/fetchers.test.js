// const fs = require('fs')
// const path = require('path')
const fetchers = require('../lib/fetchers')
const pullsFixture = require('../fixtures/github.pulls.json')
// const buildLogResultFixture = fs.readFileSync(path.resolve('fixtures/extracted.errorLog.txt'), 'utf8')

// Mock the API calls
fetchers.pulls = jest.fn().mockReturnValue(Promise.resolve(pullsFixture))
// fetchers.buildLog = jest.fn().mockReturnValue(Promise.resolve(buildLogResultFixture))

describe('fetchers', () => {
  const dummyPullsUrl = `https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls{/number}`
  const dummySha = pullsFixture[0].head.sha
  const badSha = dummySha.substr(3)
  const failedResult = fetchers.findPullBySha(dummyPullsUrl, badSha)
  const resolvedResult = fetchers.findPullBySha(dummyPullsUrl, dummySha)

  describe('.pulls()', () => {
    test('is a function', () => {
      expect(typeof fetchers.pulls).toBe(`function`)
    })

    test('returns a Promise', () => {
      expect(fetchers.pulls(dummyPullsUrl) instanceof Promise).toBe(true)
    })
  })

  describe('.findPullBySha()', () => {
    test('is a function', () => {
      expect(typeof fetchers.findPullBySha).toBe(`function`)
    })

    test('returns a Promise', () => {
      expect(resolvedResult instanceof Promise).toBe(true)
    })

    test('resolves with matching pull', (done) => {
      resolvedResult
        .then(pull => {
          expect(pull).toMatchSnapshot()
          done()
        })
    })

    test('rejects on no matching pull', () => {
      expect(failedResult).rejects.toThrow(new Error(`Pull with sha ${badSha} not found`))
    })
  })

  describe('.buildLog()', () => {
    const dummyBuildId = 398656241
    const logResult = fetchers.buildLog(dummyBuildId)

    test('is a function', () => {
      expect(typeof fetchers.buildLog).toBe(`function`)
    })

    test('returns a promise', () => {
      expect(logResult instanceof Promise).toBe(true)
    })

    test('returns a successful GET request', (done) => {
      logResult
        .then(result => {
          expect(result).toMatchSnapshot()
          done()
        })
    })

    test('rejects on missing buildId', () => {
      expect(fetchers.buildLog()).rejects.toThrow(new Error(`No buildId provided`))
    })
  })
})
