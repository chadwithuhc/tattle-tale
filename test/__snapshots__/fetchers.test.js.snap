// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchers .buildLog() returns a successful GET request 1`] = `
"$ npm test

> json-parser@1.0.0 test /home/travis/build/chadwithuhc/json-parser-code-challenge
> mocha



  Write a JSON Parser
    ✓ has a .parse() method
    ✓ supports array of numbers
    ✓ supports array of strings
    ✓ supports array of numbers and strings
    ✓ supports objects with properties
    1) supports nested objects


  5 passing (15ms)
  1 failing

  1) Write a JSON Parser
       supports nested objects:
     TypeError: Cannot read property 'trim' of undefined
      at JSONParser.autoParse (parser.js:8:15)
      at JSONParser.parseKeyValue (parser.js:79:37)
      at stripEndCharacters.split.reduce (parser.js:70:33)
      at Array.reduce (<anonymous>)
      at JSONParser.parseObject (parser.js:69:8)
      at JSONParser.autoParse (parser.js:14:21)
      at JSONParser.parseKeyValue (parser.js:79:37)
      at stripEndCharacters.split.reduce (parser.js:70:33)
      at Array.reduce (<anonymous>)
      at JSONParser.parseObject (parser.js:69:8)
      at JSONParser.autoParse (parser.js:14:21)
      at JSONParser.parse (parser.js:4:17)
      at Context.it (test.js:60:25)



npm ERR! Test failed.  See above for more details."
`;

exports[`fetchers .findPullBySha() resolves with matching pull 1`] = `
Object {
  "_links": Object {
    "comments": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/4/comments",
    },
    "commits": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4/commits",
    },
    "html": Object {
      "href": "https://github.com/chadwithuhc/json-parser-code-challenge/pull/4",
    },
    "issue": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/4",
    },
    "review_comment": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/comments{/number}",
    },
    "review_comments": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4/comments",
    },
    "self": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4",
    },
    "statuses": Object {
      "href": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/statuses/83d6439d3c48b53beaf38edaaedc70a5155b7c06",
    },
  },
  "assignee": null,
  "assignees": Array [],
  "author_association": "OWNER",
  "base": Object {
    "label": "chadwithuhc:travis-ci",
    "ref": "travis-ci",
    "repo": Object {
      "archive_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/{archive_format}{/ref}",
      "archived": false,
      "assignees_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/assignees{/user}",
      "blobs_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/blobs{/sha}",
      "branches_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/branches{/branch}",
      "clone_url": "https://github.com/chadwithuhc/json-parser-code-challenge.git",
      "collaborators_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/collaborators{/collaborator}",
      "comments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/comments{/number}",
      "commits_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/commits{/sha}",
      "compare_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/compare/{base}...{head}",
      "contents_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/contents/{+path}",
      "contributors_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/contributors",
      "created_at": "2018-05-08T02:46:53Z",
      "default_branch": "master",
      "deployments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/deployments",
      "description": "Challenge: Create a JSON Parser in JavaScript with provided tests",
      "downloads_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/downloads",
      "events_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/events",
      "fork": false,
      "forks": 0,
      "forks_count": 0,
      "forks_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/forks",
      "full_name": "chadwithuhc/json-parser-code-challenge",
      "git_commits_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/commits{/sha}",
      "git_refs_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/refs{/sha}",
      "git_tags_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/tags{/sha}",
      "git_url": "git://github.com/chadwithuhc/json-parser-code-challenge.git",
      "has_downloads": true,
      "has_issues": true,
      "has_pages": false,
      "has_projects": true,
      "has_wiki": true,
      "homepage": "",
      "hooks_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/hooks",
      "html_url": "https://github.com/chadwithuhc/json-parser-code-challenge",
      "id": 132544383,
      "issue_comment_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/comments{/number}",
      "issue_events_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/events{/number}",
      "issues_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues{/number}",
      "keys_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/keys{/key_id}",
      "labels_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/labels{/name}",
      "language": "JavaScript",
      "languages_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/languages",
      "license": null,
      "merges_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/merges",
      "milestones_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/milestones{/number}",
      "mirror_url": null,
      "name": "json-parser-code-challenge",
      "node_id": "MDEwOlJlcG9zaXRvcnkxMzI1NDQzODM=",
      "notifications_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/notifications{?since,all,participating}",
      "open_issues": 1,
      "open_issues_count": 1,
      "owner": Object {
        "avatar_url": "https://avatars2.githubusercontent.com/u/259196?v=4",
        "events_url": "https://api.github.com/users/chadwithuhc/events{/privacy}",
        "followers_url": "https://api.github.com/users/chadwithuhc/followers",
        "following_url": "https://api.github.com/users/chadwithuhc/following{/other_user}",
        "gists_url": "https://api.github.com/users/chadwithuhc/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/chadwithuhc",
        "id": 259196,
        "login": "chadwithuhc",
        "node_id": "MDQ6VXNlcjI1OTE5Ng==",
        "organizations_url": "https://api.github.com/users/chadwithuhc/orgs",
        "received_events_url": "https://api.github.com/users/chadwithuhc/received_events",
        "repos_url": "https://api.github.com/users/chadwithuhc/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/chadwithuhc/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/chadwithuhc/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/chadwithuhc",
      },
      "private": false,
      "pulls_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls{/number}",
      "pushed_at": "2018-06-30T06:22:02Z",
      "releases_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/releases{/id}",
      "size": 5,
      "ssh_url": "git@github.com:chadwithuhc/json-parser-code-challenge.git",
      "stargazers_count": 0,
      "stargazers_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/stargazers",
      "statuses_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/statuses/{sha}",
      "subscribers_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/subscribers",
      "subscription_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/subscription",
      "svn_url": "https://github.com/chadwithuhc/json-parser-code-challenge",
      "tags_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/tags",
      "teams_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/teams",
      "trees_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/trees{/sha}",
      "updated_at": "2018-06-30T01:31:23Z",
      "url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge",
      "watchers": 0,
      "watchers_count": 0,
    },
    "sha": "33808abb2f03828fae70e3b06273558b5988cd66",
    "user": Object {
      "avatar_url": "https://avatars2.githubusercontent.com/u/259196?v=4",
      "events_url": "https://api.github.com/users/chadwithuhc/events{/privacy}",
      "followers_url": "https://api.github.com/users/chadwithuhc/followers",
      "following_url": "https://api.github.com/users/chadwithuhc/following{/other_user}",
      "gists_url": "https://api.github.com/users/chadwithuhc/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/chadwithuhc",
      "id": 259196,
      "login": "chadwithuhc",
      "node_id": "MDQ6VXNlcjI1OTE5Ng==",
      "organizations_url": "https://api.github.com/users/chadwithuhc/orgs",
      "received_events_url": "https://api.github.com/users/chadwithuhc/received_events",
      "repos_url": "https://api.github.com/users/chadwithuhc/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/chadwithuhc/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/chadwithuhc/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/chadwithuhc",
    },
  },
  "body": "Long body text",
  "closed_at": null,
  "comments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/4/comments",
  "commits_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4/commits",
  "created_at": "2018-06-30T06:22:01Z",
  "diff_url": "https://github.com/chadwithuhc/json-parser-code-challenge/pull/4.diff",
  "head": Object {
    "label": "chadwithuhc:pr-test",
    "ref": "pr-test",
    "repo": Object {
      "archive_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/{archive_format}{/ref}",
      "archived": false,
      "assignees_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/assignees{/user}",
      "blobs_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/blobs{/sha}",
      "branches_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/branches{/branch}",
      "clone_url": "https://github.com/chadwithuhc/json-parser-code-challenge.git",
      "collaborators_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/collaborators{/collaborator}",
      "comments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/comments{/number}",
      "commits_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/commits{/sha}",
      "compare_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/compare/{base}...{head}",
      "contents_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/contents/{+path}",
      "contributors_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/contributors",
      "created_at": "2018-05-08T02:46:53Z",
      "default_branch": "master",
      "deployments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/deployments",
      "description": "Challenge: Create a JSON Parser in JavaScript with provided tests",
      "downloads_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/downloads",
      "events_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/events",
      "fork": false,
      "forks": 0,
      "forks_count": 0,
      "forks_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/forks",
      "full_name": "chadwithuhc/json-parser-code-challenge",
      "git_commits_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/commits{/sha}",
      "git_refs_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/refs{/sha}",
      "git_tags_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/tags{/sha}",
      "git_url": "git://github.com/chadwithuhc/json-parser-code-challenge.git",
      "has_downloads": true,
      "has_issues": true,
      "has_pages": false,
      "has_projects": true,
      "has_wiki": true,
      "homepage": "",
      "hooks_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/hooks",
      "html_url": "https://github.com/chadwithuhc/json-parser-code-challenge",
      "id": 132544383,
      "issue_comment_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/comments{/number}",
      "issue_events_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/events{/number}",
      "issues_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues{/number}",
      "keys_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/keys{/key_id}",
      "labels_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/labels{/name}",
      "language": "JavaScript",
      "languages_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/languages",
      "license": null,
      "merges_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/merges",
      "milestones_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/milestones{/number}",
      "mirror_url": null,
      "name": "json-parser-code-challenge",
      "node_id": "MDEwOlJlcG9zaXRvcnkxMzI1NDQzODM=",
      "notifications_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/notifications{?since,all,participating}",
      "open_issues": 1,
      "open_issues_count": 1,
      "owner": Object {
        "avatar_url": "https://avatars2.githubusercontent.com/u/259196?v=4",
        "events_url": "https://api.github.com/users/chadwithuhc/events{/privacy}",
        "followers_url": "https://api.github.com/users/chadwithuhc/followers",
        "following_url": "https://api.github.com/users/chadwithuhc/following{/other_user}",
        "gists_url": "https://api.github.com/users/chadwithuhc/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/chadwithuhc",
        "id": 259196,
        "login": "chadwithuhc",
        "node_id": "MDQ6VXNlcjI1OTE5Ng==",
        "organizations_url": "https://api.github.com/users/chadwithuhc/orgs",
        "received_events_url": "https://api.github.com/users/chadwithuhc/received_events",
        "repos_url": "https://api.github.com/users/chadwithuhc/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/chadwithuhc/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/chadwithuhc/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/chadwithuhc",
      },
      "private": false,
      "pulls_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls{/number}",
      "pushed_at": "2018-06-30T06:22:02Z",
      "releases_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/releases{/id}",
      "size": 5,
      "ssh_url": "git@github.com:chadwithuhc/json-parser-code-challenge.git",
      "stargazers_count": 0,
      "stargazers_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/stargazers",
      "statuses_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/statuses/{sha}",
      "subscribers_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/subscribers",
      "subscription_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/subscription",
      "svn_url": "https://github.com/chadwithuhc/json-parser-code-challenge",
      "tags_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/tags",
      "teams_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/teams",
      "trees_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/git/trees{/sha}",
      "updated_at": "2018-06-30T01:31:23Z",
      "url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge",
      "watchers": 0,
      "watchers_count": 0,
    },
    "sha": "83d6439d3c48b53beaf38edaaedc70a5155b7c06",
    "user": Object {
      "avatar_url": "https://avatars2.githubusercontent.com/u/259196?v=4",
      "events_url": "https://api.github.com/users/chadwithuhc/events{/privacy}",
      "followers_url": "https://api.github.com/users/chadwithuhc/followers",
      "following_url": "https://api.github.com/users/chadwithuhc/following{/other_user}",
      "gists_url": "https://api.github.com/users/chadwithuhc/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/chadwithuhc",
      "id": 259196,
      "login": "chadwithuhc",
      "node_id": "MDQ6VXNlcjI1OTE5Ng==",
      "organizations_url": "https://api.github.com/users/chadwithuhc/orgs",
      "received_events_url": "https://api.github.com/users/chadwithuhc/received_events",
      "repos_url": "https://api.github.com/users/chadwithuhc/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/chadwithuhc/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/chadwithuhc/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/chadwithuhc",
    },
  },
  "html_url": "https://github.com/chadwithuhc/json-parser-code-challenge/pull/4",
  "id": 198462470,
  "issue_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/issues/4",
  "labels": Array [],
  "locked": false,
  "merge_commit_sha": "38e7059ed8819afa8fc4f6765c701aaa01176cc7",
  "merged_at": null,
  "milestone": null,
  "node_id": "MDExOlB1bGxSZXF1ZXN0MTk4NDYyNDcw",
  "number": 4,
  "patch_url": "https://github.com/chadwithuhc/json-parser-code-challenge/pull/4.patch",
  "requested_reviewers": Array [],
  "requested_teams": Array [],
  "review_comment_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/comments{/number}",
  "review_comments_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4/comments",
  "state": "open",
  "statuses_url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/statuses/83d6439d3c48b53beaf38edaaedc70a5155b7c06",
  "title": "PR Test 3",
  "updated_at": "2018-06-30T16:44:30Z",
  "url": "https://api.github.com/repos/chadwithuhc/json-parser-code-challenge/pulls/4",
  "user": Object {
    "avatar_url": "https://avatars2.githubusercontent.com/u/259196?v=4",
    "events_url": "https://api.github.com/users/chadwithuhc/events{/privacy}",
    "followers_url": "https://api.github.com/users/chadwithuhc/followers",
    "following_url": "https://api.github.com/users/chadwithuhc/following{/other_user}",
    "gists_url": "https://api.github.com/users/chadwithuhc/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/chadwithuhc",
    "id": 259196,
    "login": "chadwithuhc",
    "node_id": "MDQ6VXNlcjI1OTE5Ng==",
    "organizations_url": "https://api.github.com/users/chadwithuhc/orgs",
    "received_events_url": "https://api.github.com/users/chadwithuhc/received_events",
    "repos_url": "https://api.github.com/users/chadwithuhc/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/chadwithuhc/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/chadwithuhc/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/chadwithuhc",
  },
}
`;
